src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,725 @@
+    /// mutex to become free. When this is nonzero, `wait` must be suspended until it reaches zero
+    /// again.
+    waiting_operations: AtomicUsize,
+    /// The condition variable that gets notified when `waiting_operations` reaches zero or
+    /// `notified` becomes true.
+    ///
+    /// This is used with the `fds` mutex.
+    operations_complete: Condvar,
+
+    /// This selectors id.
+    #[cfg(debug_assertions)]
+    #[allow(dead_code)]
+    id: usize,
+}
+
+/// The file descriptors to poll in a `Poller`.
+#[derive(Debug, Clone)]
+struct Fds {
+    /// The list of `pollfds` taken by poll.
+    ///
+    /// The first file descriptor is always present and is used to notify the poller.
+    poll_fds: Vec<PollFd>,
+    /// The map of each file descriptor to data associated with it. This does not include the file
+    /// descriptors created by the internal notify waker.
+    fd_data: HashMap<RawFd, FdData>,
+}
+
+/// Transparent wrapper around `libc::pollfd`, used to support `Debug` derives without adding the
+/// `extra_traits` feature of `libc`.
+#[repr(transparent)]
+#[derive(Clone)]
+struct PollFd(libc::pollfd);
+
+impl Debug for PollFd {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("pollfd")
+            .field("fd", &self.0.fd)
+            .field("events", &self.0.events)
+            .field("revents", &self.0.revents)
+            .finish()
+    }
+}
+
+/// Data associated with a file descriptor in a poller.
+#[derive(Debug, Clone)]
+struct FdData {
+    /// The index into `poll_fds` this file descriptor is.
+    poll_fds_index: usize,
+    /// The key of the `Event` associated with this file descriptor.
+    token: Token,
+    /// Used to communicate with IoSourceState when we need to internally deregister
+    /// based on a closed fd.
+    shared_record: Arc<RegistrationRecord>,
+}
+
+impl SelectorState {
+    pub fn new() -> io::Result<SelectorState> {
+        let notify_waker = WakerInternal::new()?;
+
+        Ok(Self {
+            fds: Mutex::new(Fds {
+                poll_fds: vec![PollFd(libc::pollfd {
+                    fd: notify_waker.as_raw_fd(),
+                    events: libc::POLLIN,
+                    revents: 0,
+                })],
+                fd_data: HashMap::new(),
+            }),
+            pending_removal: Mutex::new(Vec::new()),
+            pending_wake_tokens: Mutex::new(Vec::new()),
+            notify_waker,
+            waiting_operations: AtomicUsize::new(0),
+            operations_complete: Condvar::new(),
+            #[cfg(debug_assertions)]
+            id: NEXT_ID.fetch_add(1, Ordering::Relaxed),
+        })
+    }
+
+    pub fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        let deadline = timeout.map(|t| Instant::now() + t);
+
+        events.clear();
+
+        let mut fds = self.fds.lock().unwrap();
+        let mut closed_raw_fds = Vec::new();
```

Comment:
Yes, but not in a way that's all that straight forward to test.  I'm interested in this optimization as today we allocate at least once each time an fd is closed, but I avoided a potential dead lock case by trying to apply the removals while holding the fds lock.  The most efficient solution is to then move closed_raw_fds inside the fds Mutex and apply the removals while still holding the lock, but this isn't straight forward to test that we didn't introduce subtle dead locks so I'm really trying to punt here if we can.

If you feel strongly though I'll rework, though I'll really need to think hard about how to create an effective test to prevent the deadlock.

Reply:
We can pass a `close_raw_fds` vector to the function and keep it in `Selector` (not `SelectorState`).

Reply:
Oh btw, `Selector::select` can take a mutable reference to self (`&mut self`), so anything in `Selector` can be freely accessed. Maybe that wasn't clear as most Unix implementation don't make use of that.

Reply:
Normally true, but the way the poll impl works is that there needs to be a non-exclusive relationship between Registry and Selector such that `Registry::register` will ultimately call into `poll::IoSourceState::register` which accesses the selector using `registry.selector()` and then calls `register_internal`.  This test proves that we cannot have exclusive access to `poll::Selector::poll`: https://github.com/tokio-rs/mio/blob/master/tests/poll.rs#L323 (i.e. we must be able to call `SelectorState::register_internal` and `SelectorState::select` simultaneously).  This is also why we've got a Mutex for accessing `poll::SelectorState::fds` (register_internal and select both want to modify this data structure).

And that's why the dead lock can be so insidious -- if you try to re-use the memory for `closed_raw_fds` then you will have to extend the lock's duration (or add another one and there's already 3!) in a way that will start to overlap the locks on top of each other making a dead lock opportunity considerably more likely.  I'm confident I can apply the optimization without a dead lock, but I'm not so confident I can devise a reliable test that will prevent a dead lock regression from easily slipping in later.

I vote that we punt on the optimization for now given this concern, though I can push harder down this path if you prefer.

Reply:
I'm fine with punting this for now, we need to make a follow up issue any way (mostly documenting the implementation once we're ready to stabilise it). However I think you're misunderstand my suggestion.

I'm suggesting we add a field `closed_raw_fds: Vec<RawFd>` to `Selector`, change `Selector::select` to take `&mut self` allowing to accept this new field exclusively. Change `SelectorState::select` (this function) to accept `closed_raw_fds: &mut Vec<RawFd>` as an argument and use that as a buffer, rather then creating a new `Vec` in each call to `select`.

Reply:
Oh I see what you mean.  SelectorState won't hold the buffer and so won't need `&mut self`, but `Selector` can hold the buffer _however_ this means that you'll be cloning the buffer for every register that you do (see `poll::IoSourceState::register`'s use of `try_clone()`.  This means that you'll actually allocate _more_ than you were before since right now you only allocate on socket close instead of socket creation.

Reply:
Using #1701 to track these optimizations